File: src/PYB11/Solvers/NodeMap.py
from PYB11Generator import *
from MatrixMap import MatrixMap

@PYB11holder("std::shared_ptr")
@PYB11template("Dimension")
class NodeMap(MatrixMap):
    def pyinit(self,
               flatConnectivity = "FlatConnectivity<%(Dimension)s>&"):
        "Contains the indexing information for local nodes"

    @PYB11virtual
    @PYB11const
    def firstGlobalIndex(self):
        "Global index of first node"
        return "int"

    @PYB11virtual
    @PYB11const
    def lastGlobalIndex(self):
        "Global index of last node"
        return "int"

    @PYB11virtual
    @PYB11const
    def numLocalElements(self):
        "Number of elements on this processor"
        return "int"

    @PYB11virtual
    @PYB11const
    def numGlobalElements(self):
        "Number of global elements"
        return "int"

    @PYB11virtual
    @PYB11const
    def numElementsPerRow(self,
                          localIndex = "int"):
        "Get the number of nonzero columns per row"
        return "int"
    
    @PYB11virtual
    @PYB11const
    def checkClassInvariants(self):
        "Check class invariants"
        return "void"

    @PYB11virtual
    @PYB11const
    def getColumnIndices(self,
                         localRowIndex = "int",
                         globalColumnIndices = "std::vector<int>&"):
        "Get the nonzero column indices for a given row"
        return "void"
    
    @PYB11virtual
    @PYB11const
    def getLocalIndex(self,
                      nodeListIndex = "int",
                      nodeIndex = "int"):
        "Convert node indices to matrix index"
        return "int"
     
    @PYB11virtual
    @PYB11const
    def getNodeIndex(self,
                     localIndex = "int"):
        "Convert matrix index to node indices"
        return "std::pair<int, int>"

    @PYB11virtual
    @PYB11const
    def getGlobalIndex(self,
                       nodeListIndex = "int",
                       nodeIndex = "int"):
        "Convert node indices to global matrix index"
        return "int"

File: src/Solvers/NodeMap.cc
//---------------------------------Spheral++----------------------------------//
// SpheralNodeMap
//
// Creates a matrix map based on the connectivity of the Spheral nodes.
// Assumes that the global indexing is sequential on one process and
// contiguous between processors. See Utilities/globalNodeIDs.hh. 
//----------------------------------------------------------------------------//
#include "NodeMap.hh"

#include "Boundary/Boundary.hh"
#include "DataBase/DataBase.hh"
#include "Registrar/CombinedRegistrar.hh"
#include "Solvers/containsConstantNodes.hh"
#include "Utilities/DBC.hh"
#include "Utilities/globalNodeIDs.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor
//------------------------------------------------------------------------------
template<typename Dimension> 
NodeMap<Dimension>::
NodeMap(const FlatConnectivity<Dimension>& flatConnectivity):
  mConnectivity(flatConnectivity) {
}

//------------------------------------------------------------------------------
// Check that the indices make sense.
//------------------------------------------------------------------------------
template<typename Dimension>
void
NodeMap<Dimension>::
checkClassInvariants() const {
}

//------------------------------------------------------------------------------
// Get first global index belonging to this processor.
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
firstGlobalIndex() const {
  return mConnectivity.firstGlobalIndex();
}
  
//------------------------------------------------------------------------------
// Get last global index belonging to this processor.
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
lastGlobalIndex() const {
  return mConnectivity.lastGlobalIndex();
}

//------------------------------------------------------------------------------
// Get local number of nodes for this problem.
// Performs a summation over the NodeLists.
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
numLocalElements() const {
  return mConnectivity.numInternalNodes();
}

//------------------------------------------------------------------------------
// Get global number of nodes for this problem.
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
numGlobalElements() const {
  return mConnectivity.numGlobalNodes();
}

//------------------------------------------------------------------------------
// Get number of nonzero columns for this row
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
numElementsPerRow(int localRowIndex) const {
  return mConnectivity.numNonConstNeighbors(localRowIndex);
}

template<typename Dimension>
int
NodeMap<Dimension>::
numElementsPerRow(int nodeListIndex,
                  int nodeIndex) const {
  const auto localIndex = mConnectivity.nodeToLocal(nodeListIndex, nodeIndex);
  return numElementsPerRow(localIndex);
}

//------------------------------------------------------------------------------
// Get nonzero columns corresponding to this row
//------------------------------------------------------------------------------
template<typename Dimension>
void
NodeMap<Dimension>::
getColumnIndices(const int localRowIndex,
                 std::vector<int>& globalColumnIndices) const {
  mConnectivity.globalNeighborIndices(localRowIndex,
                                      globalColumnIndices);
  return;
}

//------------------------------------------------------------------------------
// Get the local matrix index for a given node index. 
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
getLocalIndex(int nodeListIndex,
              int nodeIndex) const {
  return mConnectivity.nodeToLocal(nodeListIndex, nodeIndex);
}

//------------------------------------------------------------------------------
// Get the node index for a given local index
//------------------------------------------------------------------------------
template<typename Dimension>
std::pair<int, int>
NodeMap<Dimension>::
getNodeIndex(int localIndex) const {
  return mConnectivity.localToNode(localIndex);
}

//------------------------------------------------------------------------------
// Convert local indices to global indices
//------------------------------------------------------------------------------
template<typename Dimension>
int
NodeMap<Dimension>::
getGlobalIndex(int localRowIndex) const {
  return mConnectivity.localToGlobal(localRowIndex);
}

template<typename Dimension>
int
NodeMap<Dimension>::
getGlobalIndex(int nodeListIndex,
               int nodeIndex) const {
  const auto localIndex = mConnectivity.nodeToLocal(nodeListIndex, nodeIndex);
  return mConnectivity.localToGlobal(localIndex);
}
//------------------------------------------------------------------------------
// Return whether node is a constant boundary
//------------------------------------------------------------------------------
template<typename Dimension>
bool 
NodeMap<Dimension>::
isConstantBoundaryNode(int nodeListIndex,
                       int nodeIndex) const {
  const auto localIndex = mConnectivity.nodeToLocal(nodeListIndex, nodeIndex);
  return mConnectivity.isConstantBoundaryNode(localIndex);
}

} // end namespace Spheral
